Fix parsing of OCaml code by ocamlyacc

DemiMarie opened the pull request:
ocamlyacc would missparse certain valid OCaml code.  As shown in [MPR#7454] and [MPR#7455], this can cause valid `.mly` files to be rejected, and for certain invalid files to be compiled to OCaml code that segfaults, without the `.mly` file containing any unsafe functions.

This patch fixes both problems.

[MPR#7454]: https://caml.inria.fr/mantis/view.php?id=7454
[MPR#7455]: https://caml.inria.fr/mantis/view.php?id=7455

dra27 commented on yacc/closure.c:
```diff
@@ -110,8 +110,9 @@ void set_first_derives(void)
-
+#ifdef DEBUG
       vrow += varsetsize;
+#endif
```
What is this change doing?

DemiMarie replied:
The Clang static analyzer said that the store on line 114 is dead.

dra27 replied:
So it is - but in which case the line can be deleted (in the tidying up commit)...

dra27 commented on yacc/main.c:
```diff
@@ -138,7 +138,7 @@ void done(int k)
 void onintr(int dummy)
 {
-    done(1);
+    done(128 + dummy);
```
Similarly - what's this alteration for?

DemiMarie replied:
Unix convention is that signals cause the program to exit with status 128 + the signal number.

dra27 replied:
That may well be so, but that's definitely not something to be smuggled inside another patch. Could this change at least be put on its own (I'm not entirely convinced that that belongs in this pull request at all)

dra27 commented on yacc/reader.c:
```diff
@@ -497,126 +619,6 @@ void copy_text(void)
-    }
-}
-
-
```
Is eliminating this part of the patch, or just tidying?

DemiMarie replied:
It is tidying.

dra27 commented:
Thanks for the patch - a few things, though, please. There are various things which look unrelated to the two PRs (sorry if I'm misreading the code) - this single commit seems to combine some tidying, some necessary refactoring and fixing two PRs. I think I'm correct that fixing the actual code changes for the two PRs are not directly related, so please could you split it into four commits to create a slightly more bisect-able story:

1. Tidying any superfluous code (is that just a legacy of its original porting from yacc?)
2. Refactoring: the moving of caml_ident_start, etc. and moving **unchanged** bodies of code out to process_quoted_string, etc.
3. Fix one PR
4. Fix the other PR

Please also do a Changes entry (again, these aren't related so it should be really be two entries - it's fine fixing in one pull request, of course) and if you could bear to turn the two sample files you sent into testsuite tests, that would be great (I can do that afterwards if you don't have time/inclination).

DemiMarie committed:
Delete some superfluous code

The %ident and %union directives don't make sense for OCaml,
and generated broken OCaml code.  Just drop support for them.

DemiMarie committed:
Refactor the ocamlyacc reader

The processing of apostrophes, quoted strings and comments was
duplicated in at least two places.  This refactors both into
functions.

Also move the bitmap code, for use by later changes.

DemiMarie committed:
Fix parsing of raw strings

This adds support for parsing of raw OCaml strings.

DemiMarie committed:
Check for balanced parentheses and curly braces in ocamlyacc

This uses an explicit stack to keep track of parentheses and curly
braces in actions.

DemiMarie committed:
Fix the parsing of comments by ocamlyacc

DemiMarie committed:
Fix parsing for a corner case in ocamlyacc

A'"'" was misparsed (the ' was not associated with the A as it should
be)

DemiMarie committed:
Partial testcase and bug fix

This adds a partial testcase for the better lexing of
OCaml code by ocamlyacc, as well as a bug fix.

DemiMarie committed:
Delete code that isn't useful for ocamlyacc

It was left from the conversion of Berkely Yacc to ocamlyacc

DemiMarie committed:
Add Changes entry

DemiMarie committed:
Fix build on MSVC

DemiMarie commented:
@dra27 done.

dra27 commented:
Sorry for the delay looking at this - many thanks for the refactoring and the extra parts! I'm just running it through the full CI at Inria, as that includes a few other C compilers, but I propose merging once that's complete.

dra27 commented:
OK, the failures in precheck on mingw32/mingw64 are unrelated, so merging.

gasche commented:
In the Changes file, this change is marked as compatibility-breaking (use of the `*` bullet instead of `-`). @DemiMarie, how does it break compatibility (besides fixing wrong behaviors)?

DemiMarie commented:
It doesn't.

On Jul 12, 2017 10:45 AM, "Gabriel Scherer" <notifications@github.com>
wrote:

> In the Changes file, this change is marked as compatibility-breaking (use
> of the * bullet instead of -). @DemiMarie <https://github.com/demimarie>,
> how does it break compatibility (besides fixing wrong behaviors)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ocaml/ocaml/pull/1012#issuecomment-314791950>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGGWB8Y9Fev7MKIFneVpkSE9UfSrMc5Jks5sNNvhgaJpZM4LjzmX>
> .
>

gasche commented:
Thanks, I fixed the Changes in #1241.
